### PR TITLE
chainHead: Track reported blocks to capture notification gaps

### PR DESCRIPTION
**Pull request by lexnv**
There are cases during warp sync or re-orgs, where we receive a notification with a block parent that was not reported in the past. This PR extends the tracking state to catch those cases and report a `Stop` event to the user.

This PR adds a new state to the RPC-v2 chainHead to track which blocks have been reported.

In the past we relied on the pinning mechanism to provide us details if a block is pinned or not.
However, the pinning state keeps the minimal information around for pinning. Therefore, unpinning a block will cause the state to disappear.

Closes: https://github.com/paritytech/polkadot-sdk/issues/5761

### CODE REVIEW

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 1]
```diff
@@ -71,6 +69,8 @@ pub struct ChainHeadFollower<BE: Backend<Block>, Block: BlockT, Client> {
 	current_best_block: Option<Block::Hash>,
 	/// LRU cache of pruned blocks.
 	pruned_blocks: LruMap<Block::Hash, ()>,
```
Comment (lexnv) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:71]: Believe that `pruned_blocks` are no longer needed (https://github.com/paritytech/polkadot-sdk/issues/5605), will tackle this in another PR

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 2]
```diff
@@ -71,6 +69,8 @@ pub struct ChainHeadFollower<BE: Backend<Block>, Block: BlockT, Client> {
...
+	/// LRU cache of pruned blocks.
```
Comment (skunert) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:72]:
```suggestion
	/// LRU cache of announced blocks.
```

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 3]
```diff
@@ -96,6 +96,9 @@ impl<BE: Backend<Block>, Block: BlockT, Client> ChainHeadFollower<BE, Block, Cli
 			pruned_blocks: LruMap::new(ByLength::new(
 				MAX_PINNED_BLOCKS.try_into().unwrap_or(u32::MAX),
 			)),
+			announced_blocks: LruMap::new(ByLength::new(
```
Comment (skunert) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:99]: One downside of this approach is that we never remove anything manually and just rely on the LRU to remove items. However, that means that over time we will have the maximum amount of items in every LRU for every subscription. But in reality probably does not matter, for these 512 items thats approx. 16kb of memory usage then :shrug:

**Review by skunert — APPROVED** [review]

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 4]
```diff
@@ -71,11 +69,56 @@ pub struct ChainHeadFollower<BE: Backend<Block>, Block: BlockT, Client> {
...
+			self.blocks.insert(block, ());
+		}
+	}
+
+	/// Check if the block was previously announced.
+	fn was_announced(&mut self, block: &Block::Hash) -> bool {
+		// For the finalized blocks we use `peek` to avoid moving the block counter to the front.
+		// This effectively means that the LRU acts as a FIFO queue. Otherwise, we might
+		// end up with scenarios where the last inserted finalized block is removed.
+		self.blocks.get(block).is_some() || self.finalized.peek(block).is_some()
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:118]: I think it's quite "strange/hard to understand" use a LRU for the finalized blocks when it's treated as a FIFO queue.

Personally I would just use a fixed size ring buffer but probably less efficient to do lookups.

An alternative is create a wrapper type such as:

```rust
/// A wrapper type to utilize an LruMap as FIFO queue.
struct FinalizedBlocksFifo(LruMap<Block::Hash, ()>);

impl FinalizedBlocksFifo {
    fn insert(&mut self, block: Block::Hash) {
        self.finalized.insert(block, ());
     }

    fn contains(&mut self, block: &Block::Hash) -> bool {
        self.0.peek(block).is_some()
    }
}
```

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 5]
```diff
@@ -71,11 +69,56 @@ pub struct ChainHeadFollower<BE: Backend<Block>, Block: BlockT, Client> {
...
+		} else {
+			self.blocks.insert(block, ());
+		}
+	}
+
+	/// Check if the block was previously announced.
+	fn was_announced(&mut self, block: &Block::Hash) -> bool {
+		// For the finalized blocks we use `peek` to avoid moving the block counter to the front.
+		// This effectively means that the LRU acts as a FIFO queue. Otherwise, we might
+		// end up with scenarios where the last inserted finalized block is removed.
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:117]:
```suggestion
		// end up with scenarios where the "finalized block" is the end of LRU is overwritten which may not necessarily be the oldest finalized block i.e, possible that get promotes an older finalized block but it was accessed more recentely 
```

**Review by niklasad1 — APPROVED** [review]

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 6]
```diff
@@ -71,11 +69,75 @@ pub struct ChainHeadFollower<BE: Backend<Block>, Block: BlockT, Client> {
...
+
+impl<Block: BlockT> MostRecentFinalizedBlocks<Block> {
+	/// Insert the finalized block hash into the LRU cache.
+	fn insert(&mut self, block: Block::Hash) {
+		self.0.insert(block, ());
+	}
+
+	/// Check if the block is contained in the LRU cache.
+	fn contains(&mut self, block: &Block::Hash) -> Option<&()> {
+		// For the finalized blocks we use `peek` to avoid moving the block counter to the front.
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:97]: I think it's sufficient with the comments on the struct itself to explain that the LRU is used as FIFO queue but maybe move some of the comment here to struct itself?

but up to you :)

Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs]: Thanks Niklas! That sounds good to me! 🙏

**Pull request merged**